### ISSUE TITLE
OMP warning and segfault on Mac OS 14.0

### ISSUE THREAD

**Issue by anirban-mu** [MaartenGr/BERTopic#1580]
Hi all,

I am getting the following OMP warning and segfault on a "fresh" install of Python 3 and BERTopic. 

Specifically, I deleted all site packages in my old Python 3.11.6 install, uninstalled Python 3, and reinstalled Python3 using HomeBrew.

Python version: 3.11.6
Mac OS Sonoma 14.0

Then I installed BERTopic using pip. I did not install anything else.

I get the following warning and error regardless of the number of documents passed to BERTopic. I saw in some places that the issue could be numba. I am on numba 0.58.0, which is the latest version. I tried forcing numba 0.57.1 but that did not help. 

This occurs even when using the default code but I am presenting the complete code in the hope that it may be useful to someone who has a guess as to the issue.

Code:

        from umap import UMAP
        from hdbscan import HDBSCAN
        from sentence_transformers import SentenceTransformer
        from sklearn.feature_extraction.text import CountVectorizer
        
        from bertopic import BERTopic
        from bertopic.vectorizers import ClassTfidfTransformer
        from bertopic.representation import KeyBERTInspired
        
        # Use the representation model in BERTopic on top of the default pipeline
        topic_model = BERTopic(representation_model=representation_model)
        
        # Step 1 - Extract embeddings
        embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
        
        # Step 2 - Reduce dimensionality
        umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine')
        
        # Step 3 - Cluster reduced embeddings
        hdbscan_model = HDBSCAN(min_cluster_size=15, metric='euclidean', cluster_selection_method='eom', prediction_data=True)
        
        # Step 4 - Tokenize topics
        vectorizer_model = CountVectorizer(stop_words="english")
        
        # Step 5 - Create topic representation
        ctfidf_model = ClassTfidfTransformer()
        
        # Step 6 - (Optional) Fine-tune topic representations with 
        # a `bertopic.representation` model
        representation_model = KeyBERTInspired()
        
        # All steps together
        topic_model = BERTopic(
          embedding_model=embedding_model,          # Step 1 - Extract embeddings
          umap_model=umap_model,                    # Step 2 - Reduce dimensionality
          hdbscan_model=hdbscan_model,              # Step 3 - Cluster reduced embeddings
          vectorizer_model=vectorizer_model,        # Step 4 - Tokenize topics
          ctfidf_model=ctfidf_model,                # Step 5 - Extract topic words
          representation_model=representation_model # Step 6 - (Optional) Fine-tune topic represenations
        )
        
        from sklearn.datasets import fetch_20newsgroups
        
        docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']
        
        topic_model.fit_transform(docs)

Warning and segfault:

        >>> topic_model.fit_transform(docs)
        OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
        zsh: segmentation fault  python3
        anirbanmukherjee@NCC-1701 python3.11 % /usr/local/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
          warnings.warn('resource_tracker: There appear to be %d '

Python packages as reported by pip:

Package               Version
--------------------- ------------
aiohttp               3.8.6
aiosignal             1.3.1
async-timeout         4.0.3
attrs                 23.1.0
bertopic              0.15.0
certifi               2023.7.22
charset-normalizer    3.3.0
click                 8.1.7
Cython                0.29.36
filelock              3.12.4
frozenlist            1.4.0
fsspec                2023.9.2
hdbscan               0.8.33
huggingface-hub       0.17.3
idna                  3.4
Jinja2                3.1.2
joblib                1.3.2
llvmlite              0.41.0
MarkupSafe            2.1.3
mpmath                1.3.0
multidict             6.0.4
networkx              3.1
nltk                  3.8.1
numba                 0.58.0
numpy                 1.25.2
openai                0.28.1
packaging             23.2
pandas                2.1.1
Pillow                10.1.0
pip                   23.2.1
plotly                5.17.0
pynndescent           0.5.10
python-dateutil       2.8.2
pytz                  2023.3.post1
PyYAML                6.0.1
regex                 2023.10.3
requests              2.31.0
safetensors           0.4.0
scikit-learn          1.3.1
scipy                 1.11.3
sentence-transformers 2.2.2
sentencepiece         0.1.99
setuptools            68.2.2
six                   1.16.0
sympy                 1.12
tbb                   2021.10.0
tenacity              8.2.3
threadpoolctl         3.2.0
tokenizers            0.14.1
torch                 2.1.0
torchvision           0.16.0
tqdm                  4.66.1
transformers          4.34.0
typing_extensions     4.8.0
tzdata                2023.3
umap-learn            0.5.4
urllib3               2.0.7
wheel                 0.41.2
yarl                  1.9.2

Any ideas on what could be the issue? Would deeply appreciate any suggestions, thanks!

**Comment by anirban-mu** [comment]
This works.

        import numba
        
        # Same code as earlier 

        numba.set_num_threads(1) # Force single threaded operation
        topic_model.fit_transform(docs[0:999])

One still gets the warning `OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.` but the code runs and there is no segfault.

There is a performance hit from turning off OMP. Hence, fit is called only on 1000 documents in the MWE. But this is ok for my current application as the dataset is not that large, so I will move on. 

If you have specific ideas I can try to help debug but this could be very hard to fix given the interaction between BERTopic, HomeBrew, dependencies, numba, and umap -- I have no idea where the issue is hiding.  Therefore, closing the bug report as I am unsure of what else we could do this far upstream from the underlying issue.

**Comment by MaartenGr** [comment]
Thanks for sharing both the issue as well as a preliminary solution! It is indeed not an ideal solution and I am not sure what triggered the issue. Hopefully, someone else can pitch in or share their experiences if they experienced the same issue.

**Comment by anirban-mu** [comment]
Yep, you bet! Apple Clang (LLVM) does not support OpenMP. Anything that calls OpenMP ends up using the HomeBrew toolchain (or equivalent non Apple install/libraries). That causes heaps of trouble always because even though the HomeBrew folks try, it’s hard to keep the toolchain in sync.

I understand why many people may advocate moving to Linux as a function of such issues but that’s a much more involved discussion. I only mention the above to apprise you that I would imagine this issue may persist unfortunately.
